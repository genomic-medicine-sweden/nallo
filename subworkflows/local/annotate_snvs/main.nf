include { ANNOTATE_CADD                               } from '../annotate_cadd/main'
include { BCFTOOLS_FILLTAGS as BCFTOOLS_FILLTAGS_ANNO } from '../../../modules/local/bcftools/filltags/main'
include { BCFTOOLS_VIEW                               } from '../../../modules/nf-core/bcftools/view/main'
include { ECHTVAR_ANNO                                } from '../../../modules/local/echtvar/anno/main'
include { ENSEMBLVEP_VEP as ENSEMBLVEP_SNV            } from '../../../modules/nf-core/ensemblvep/vep/main'
include { TABIX_TABIX as TABIX_ENSEMBLVEP_SNV         } from '../../../modules/nf-core/tabix/tabix/main'

workflow ANNOTATE_SNVS {

    take:
    ch_vcf                    // channel: [mandatory] [ val(meta), path(vcf) ]
    ch_databases              // channel:  [optional] [ path(db) ]
    ch_fasta                  // channel: [mandatory] [ val(meta), path(fasta) ]
    ch_fai                    // channel: [mandatory] [ val(meta), path(fai) ]
    ch_vep_cache              // channel: [mandatory] [ path(cache) ]
    val_vep_cache_version     //  string: [mandatory] default: 110
    ch_vep_extra_files        // channel: [mandatory] [ path(files) ]
    annotate_cadd             //    bool: [mandatory] should CADD be used to annotate indels
    annotate_echtvar          //    bool: [mandatory] should echtvar be used to annotate variants
    ch_cadd_header            // channel:  [optional] [ path(txt) ]
    ch_cadd_resources         // channel:  [optional] [ val(meta), path(annotation) ]
    ch_cadd_prescored_indels  // channel:  [optional] [ val(meta), path(prescored) ]
    pre_vep_filter            //    bool: [mandatory] should filtering be done before annotating with CADD and VEP

    main:
    ch_versions = Channel.empty()

    // Annotate with chosen databases
    if (annotate_echtvar) {
        ECHTVAR_ANNO (
            ch_vcf,
            ch_databases
        )
        ch_versions = ch_versions.mix(ECHTVAR_ANNO.out.versions)
    }

    // Allows for filtering before annotating with VEP
    if (pre_vep_filter) {
        (annotate_echtvar ? ECHTVAR_ANNO.out.bcf : ch_vcf)
            .map { meta, vcf -> [ meta, vcf, [] ] }
            .set { ch_bcftools_view_input }

        BCFTOOLS_VIEW (
            ch_bcftools_view_input,
            [],
            [],
            []
        )
        ch_versions = ch_versions.mix(BCFTOOLS_VIEW.out.versions)
    }

    //
    // Annotating with CADD
    //
    // If we don't have an index the subworkflow will create one, but needs to join on meta to create a [ meta, vcf, []Â ] channel
    if (annotate_cadd) {
        ANNOTATE_CADD (
            ch_fai,
            pre_vep_filter ? BCFTOOLS_VIEW.out.vcf : annotate_echtvar ? ECHTVAR_ANNO.out.bcf : ch_vcf,
            pre_vep_filter ? BCFTOOLS_VIEW.out.tbi : ch_vcf.map { meta, _vcf -> [ meta, [] ] },
            ch_cadd_header,
            ch_cadd_resources,
            ch_cadd_prescored_indels
        )
        ch_versions = ch_versions.mix(ANNOTATE_CADD.out.versions)
    }

    (annotate_cadd ? ANNOTATE_CADD.out.vcf : pre_vep_filter ? BCFTOOLS_VIEW.out.vcf : annotate_echtvar ? ECHTVAR_ANNO.out.bcf : ch_vcf)
        .map { meta, vcf -> [ meta, vcf, [] ] }
        .set { ch_vep_in }

    // Always annotate with VEP
    ENSEMBLVEP_SNV (
        ch_vep_in,
        "GRCh38",
        "homo_sapiens",
        val_vep_cache_version,
        ch_vep_cache,
        ch_fasta,
        ch_vep_extra_files
    )
    ch_versions = ch_versions.mix(ENSEMBLVEP_SNV.out.versions)

    TABIX_ENSEMBLVEP_SNV (
        ENSEMBLVEP_SNV.out.vcf
    )
    ch_versions = ch_versions.mix(TABIX_ENSEMBLVEP_SNV.out.versions)

    emit:
    vcf      = ENSEMBLVEP_SNV.out.vcf
    tbi      = TABIX_ENSEMBLVEP_SNV.out.tbi
    versions = ch_versions
}
