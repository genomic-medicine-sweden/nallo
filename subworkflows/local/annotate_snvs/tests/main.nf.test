nextflow_workflow {

    name "Test Workflow ANNOTATE_SNVS"
    script "../main.nf"
    workflow "ANNOTATE_SNVS"
    config "./nextflow.config"

    setup {
        run("GUNZIP") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id:'hg38' ],
                    file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                ]
                """
            }
        }
        run("SAMTOOLS_FAIDX") {
            script "../../../../modules/nf-core/samtools/faidx/main.nf"
            process {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = [[],[]]
                input[2] = false
                """
            }
        }
        run("UNTAR") {
            script "../../../../modules/nf-core/untar/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'vep_cache' ],
                    file(params.pipelines_testdata_base_path + 'reference/vep_cache_test_data.tar.gz', checkIfExists:true)
                ]
                """
            }
        }

    }

    test("bcf, db, vep_cache, '110'") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/NIST_deepvariant_norm.vcf.gz', checkIfExists: true)
                ])
                input[1] = [
                    file(params.pipelines_testdata_base_path + 'reference/cadd.v1.6.hg38.test_data.zip', checkIfExists: true)
                ]
                input[2] = GUNZIP.out.gunzip
                input[3] = SAMTOOLS_FAIDX.out.fai
                input[4] = UNTAR.out.untar.map { meta, cache -> cache }
                input[5] = Channel.value('110')
                input[6] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ]).splitCsv(header:true).map { row -> row.vep_files }.collect()
                input[7] = false
                input[8] = true
                input[9] = Channel.value([])
                input[10] = null
                input[11] = null
                input[12] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    file(workflow.out.tbi.get(0).get(1)).name,
                    path(workflow.out.vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.vcf.get(0).get(1)).vcf.summary,
                ).match() }
            )
        }

    }

    test("bcf, db, vep_cache, '110', filter before vep") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/NIST_deepvariant_norm.vcf.gz', checkIfExists: true)
                ])
                input[1] = [
                    file(params.pipelines_testdata_base_path + 'reference/cadd.v1.6.hg38.test_data.zip', checkIfExists: true)
                ]
                input[2] = GUNZIP.out.gunzip
                input[3] = SAMTOOLS_FAIDX.out.fai
                input[4] = UNTAR.out.untar.map { meta, cache -> cache }
                input[5] = Channel.value('110')
                input[6] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ]).splitCsv(header:true).map { row -> row.vep_files }.collect()
                input[7] = false
                input[8] = true
                input[9] = Channel.value([])
                input[10] = null
                input[11] = null
                input[12] = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    file(workflow.out.tbi.get(0).get(1)).name,
                    path(workflow.out.vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.vcf.get(0).get(1)).vcf.summary,
                ).match() }
            )
        }

    }

    test("bcf, [], vep_cache, '110' - no echtvar annotation") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/NIST_deepvariant_norm.vcf.gz', checkIfExists: true)
                ])
                input[1] = Channel.value([])
                input[2] = GUNZIP.out.gunzip
                input[3] = SAMTOOLS_FAIDX.out.fai
                input[4] = UNTAR.out.untar.map { meta, cache -> cache }
                input[5] = Channel.value('110')
                input[6] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ]).splitCsv(header:true).map { row -> row.vep_files }.collect()
                input[7] = false
                input[8] = false
                input[9] = Channel.value([])
                input[10] = null
                input[11] = null
                input[12] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    file(workflow.out.tbi.get(0).get(1)).name,
                    path(workflow.out.vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.vcf.get(0).get(1)).vcf.summary,
                ).match() }
            )
        }

    }

    test("bcf, [], vep_cache, '110' - no echtvar annotation - filter before vep") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/NIST_deepvariant_norm.vcf.gz', checkIfExists: true)
                ])
                input[1] = Channel.value([])
                input[2] = GUNZIP.out.gunzip
                input[3] = SAMTOOLS_FAIDX.out.fai
                input[4] = UNTAR.out.untar.map { meta, cache -> cache }
                input[5] = Channel.value('110')
                input[6] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ]).splitCsv(header:true).map { row -> row.vep_files }.collect()
                input[7] = false
                input[8] = false
                input[9] = Channel.value([])
                input[10] = null
                input[11] = null
                input[12] = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    file(workflow.out.tbi.get(0).get(1)).name,
                    path(workflow.out.vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.vcf.get(0).get(1)).vcf.summary,
                ).match() }
            )
        }

    }

    test("bcf, db, vep_cache, '110', -stub") {

        options "-stub"

        when {
            params {
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/NIST_deepvariant_norm.vcf.gz', checkIfExists: true)
                ])
                input[1] = [
                    file(params.pipelines_testdata_base_path + 'reference/cadd.v1.6.hg38.test_data.zip', checkIfExists: true)
                ]
                input[2] = GUNZIP.out.gunzip
                input[3] = SAMTOOLS_FAIDX.out.fai
                input[4] = UNTAR.out.untar.map { meta, cache -> cache }
                input[5] = Channel.value('110')
                input[6] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ]).splitCsv(header:true).map { row -> row.vep_files }.collect()
                input[7] = false
                input[8] = true
                input[9] = Channel.value([])
                input[10] = null
                input[11] = null
                input[12] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    workflow.out.tbi,
                    workflow.out.vcf
                ).match() }
            )
        }

    }

    test("bcf, db, vep_cache, '110', -stub with CADD") {

        options "-stub"

        when {
            params {
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'testdata/NIST_deepvariant_norm.vcf.gz', checkIfExists: true)
                ])
                input[1] = [
                    file(params.pipelines_testdata_base_path + 'reference/cadd.v1.6.hg38.test_data.zip', checkIfExists: true)
                ]
                input[2] = GUNZIP.out.gunzip
                input[3] = SAMTOOLS_FAIDX.out.fai
                input[4] = UNTAR.out.untar.map { meta, cache -> cache }
                input[5] = Channel.value('110')
                input[6] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ]).splitCsv(header:true).map { row -> row.vep_files }.collect()
                input[7] = false
                input[8] = true
                input[9] = Channel.of([
                    [ id: 'test' ],
                    file("$projectDir/assets/cadd_to_vcf_header_-1.0-.txt", checkIfExists: true)
                ])
                input[10] = Channel.of([
                    [ id: 'test' ],
                    file("$projectDir/assets/", checkIfExists: true)
                ])
                input[11] = Channel.of([
                    [ id: 'test' ],
                    file("$projectDir/docs/", checkIfExists: true)
                ])
                input[12] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    workflow.out.tbi,
                    workflow.out.vcf
                ).match() }
            )
        }

    }

}
