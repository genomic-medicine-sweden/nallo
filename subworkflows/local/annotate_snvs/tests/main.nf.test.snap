{
    "bcf, db, vep_cache, '110', -stub with CADD": {
        "content": [
            [
                "versions.yml:md5,6bdd1355c6de060367f50ffb5680d786",
                "versions.yml:md5,81c3ff9e5e59ed617d4e1f20ffb3f72a",
                "versions.yml:md5,8da1323d9b58551409825bee6d7d78bd"
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-28T11:47:33.617852704"
    },
    "bcf, db, vep_cache, '110'": {
        "content": [
            [
                "versions.yml:md5,6bdd1355c6de060367f50ffb5680d786",
                "versions.yml:md5,81c3ff9e5e59ed617d4e1f20ffb3f72a",
                "versions.yml:md5,8da1323d9b58551409825bee6d7d78bd"
            ],
            "test.vcf.gz.tbi",
            "86a258992d944b13f9ea2308bfefa060",
            "VcfFile [chromosomes=[chrX, chr16, chr17], sampleCount=1, variantCount=109, phased=false, phasedAutodetect=false]"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-30T11:18:27.899635439"
    },
    "bcf, db, vep_cache, '110', -stub": {
        "content": [
            [
                "versions.yml:md5,6bdd1355c6de060367f50ffb5680d786",
                "versions.yml:md5,81c3ff9e5e59ed617d4e1f20ffb3f72a",
                "versions.yml:md5,8da1323d9b58551409825bee6d7d78bd"
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-28T17:04:20.889907512"
    },
    "bcf, [], vep_cache, '110' - no echtvar annotation - filter before vep": {
        "content": [
            [
                "versions.yml:md5,81c3ff9e5e59ed617d4e1f20ffb3f72a",
                "versions.yml:md5,8da1323d9b58551409825bee6d7d78bd",
                "versions.yml:md5,fe6d181bee052fa9431323e74c5ae476"
            ],
            "test.vcf.gz.tbi",
            "61bcc566deafdbcc49dc4f1eacfa397a",
            "VcfFile [chromosomes=[chrX, chr16], sampleCount=1, variantCount=11, phased=false, phasedAutodetect=false]"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-30T11:23:32.874203087"
    },
    "bcf, db, vep_cache, '110', filter before vep": {
        "content": [
            [
                "versions.yml:md5,6bdd1355c6de060367f50ffb5680d786",
                "versions.yml:md5,81c3ff9e5e59ed617d4e1f20ffb3f72a",
                "versions.yml:md5,8da1323d9b58551409825bee6d7d78bd",
                "versions.yml:md5,fe6d181bee052fa9431323e74c5ae476"
            ],
            "test.vcf.gz.tbi",
            "29e0e8c68b142cd1333013b20cb36a74",
            "VcfFile [chromosomes=[chrX, chr16], sampleCount=1, variantCount=11, phased=false, phasedAutodetect=false]"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-30T11:20:46.84126129"
    },
    "bcf, [], vep_cache, '110' - no echtvar annotation": {
        "content": [
            [
                "versions.yml:md5,81c3ff9e5e59ed617d4e1f20ffb3f72a",
                "versions.yml:md5,8da1323d9b58551409825bee6d7d78bd"
            ],
            "test.vcf.gz.tbi",
            "59756676e6aa8832a42657407eda339d",
            "VcfFile [chromosomes=[chrX, chr16, chr17], sampleCount=1, variantCount=109, phased=false, phasedAutodetect=false]"
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-07-30T11:21:16.158860453"
    }
}